### ISSUE TITLE
Config values marked as `Unused` cause startup to fail if passed through command-line

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylla#16338]
Let's take the recently marked-as-Unused `force_schema_commit_log` option (c8ee7d4499cb5a15d3e66b51f2e8b6ee24e03791)

If we add it back to conf/scylla.yaml and start Scylla, it's ignored as expected. Even not a warning is printed, that is indeed what value_status::Unused` is for. This logic in utils/config_file.cc handles it:
```cpp
        switch (cfg.status()) {
        case value_status::Deprecated:
            h(label, "Option is deprecated", cfg.status());
            continue;
        case value_status::Invalid:
            h(label, "Option is not applicable", cfg.status());
            continue;
        case value_status::Unused:
        default:
            break;
        }
```

(`h` throws an error, and it's not called for `Unused` values).

This is how it should work, so that when we remove parameters, existing installations with possibly outdated scylla.yamls are not broken.

However, if we pass it through command line:
`--force-schema-commit-log true`

It causes Scylla to fail on startup.

This might break installations which start Scylla with custom commandline params.

**Comment by kbr-scylla** [comment]
cc @avikivity @elcallio

**Comment by nyh** [comment]
I asked the same question in my review of the recent Unused/Deprecated change - why we need different error behavior if the option came from the command line or the yaml file.

There is apparently some sort of assumption on who and why uses scylla.yaml and who and why uses command line parameters, but I'm not sure what this assumption is based on. In my own little world, in test/cql-pytest/run.py I decided to use command line parameters because I like this Unix interface better than Yaml. But IIRC, test/topology uses yaml or even both yaml and command line parameters. So these two tools use command line parameters and yaml interchangably. So I don't think the error handling in both approaches should differ.

**Comment by kbr-scylla** [comment]
@ptrsmrn 

> force_schema_commit_log should not be removed from the config options, but rather marked with value_status::Deprecated

Why? The usual meaning of Deprecated, IIUC, is that the option still has meaning and changes the behavior, but it's not recommended to be used any more and will be removed in future versions.

`force_schema_commit_log` is actually **Unused** after https://github.com/scylladb/scylladb/commit/c8ee7d4499cb5a15d3e66b51f2e8b6ee24e03791. In the sense that Scylla completely ignores whether it's true or false and does its own thing.

**Comment by ptrsmrn** [comment]
@kbr-scylla since `value_status::Deprecated` introduction, the process of removing config options should involve Deprecation step (later marking a config option Unused).

**Comment by kbr-scylla** [comment]
I guess it depends on the config option, for some options it may not make sense to have a deprecation period, and `force_schema_commit_log` is one such option.

**Comment by kbr-scylla** [comment]
The reason is that once you've started your cluster in 5.2 (or later version) at least once, this config option no longer does anything.

So in 5.5/later the only sensible thing to do is to remove it.

**Comment by ptrsmrn** [comment]
From what you say I understand that Unused/Invalid is not a way to go for this option, because setups using it would crash on startup, but neither Deprecated is. Nevertheless, I'd leave it as Deprecated, and not do anything if this option is provided, we already have a precedence for such this, see https://github.com/scylladb/scylladb/blob/master/db/config.cc#L354

**Comment by kbr-scylla** [comment]
> because setups using it would crash on startup

only if passed from command-line, which I think is a bug (this issue).

**Comment by ptrsmrn** [comment]
> > because setups using it would crash on startup
> 
> only if passed from command-line, which I think is a bug (this issue).

Yes, this has been identified by some as something we don't really like, if not a bug by others, yet we won't fix it. It's been already discussed, please see especially https://github.com/scylladb/scylladb/pull/16184 and attached google doc.

**Comment by kbr-scylla** [comment]
> yet we won't fix it

I couldn't find the explanation for that -- why not fix it?

Treating options from cmdline different than those from config file is just an oversight of the implementation, whether or not we introduce a separate "Deprecated" state.

> Nevertheless, I'd leave it as Deprecated, and not do anything if this option is provided

If I interpret the code correctly, it **throws an error** if a Deprecated option is provided; so leaving it as Deprecated contradicts with "not do anything if this option is provided"
```cpp
        switch (cfg.status()) {
        case value_status::Deprecated:
            h(label, "Option is deprecated", cfg.status());
            continue;
        case value_status::Invalid:
            h(label, "Option is not applicable", cfg.status());
            continue;
        case value_status::Unused:
        default:
            break;
        }
```

only the `Unused` case does "not do anything". Except if it's provided by command-line -- that last part is inconsistent behavior.

**Comment by ptrsmrn** [comment]
> > yet we won't fix it
> 
> I couldn't find the explanation for that -- why not fix it?

I don't remember, probably doesn't have any impact on anything, hence is a low priority, also possibly works that way due to historical reasons.

> Treating options from cmdline different than those from config file is just an oversight of the implementation, whether or not we introduce a separate "Deprecated" state.

Right now, when we have `value_status::Deprecated`, that every deprecated config option should go through, it doesn't matter, I believe.

> > Nevertheless, I'd leave it as Deprecated, and not do anything if this option is provided
> 
> If I interpret the code correctly, it **throws an error** if a Deprecated option is provided; so leaving it as Deprecated contradicts with "not do anything if this option is provided"
> 
> ```c++
>         switch (cfg.status()) {
>         case value_status::Deprecated:
>             h(label, "Option is deprecated", cfg.status());
>             continue;
>         case value_status::Invalid:
>             h(label, "Option is not applicable", cfg.status());
>             continue;
>         case value_status::Unused:
>         default:
>             break;
>         }
> ```
> 
> only the `Unused` case does "not do anything". Except if it's provided by command-line -- that last part is inconsistent behavior.

I built OSS on: 
```
[piotrs@fedora scylla2]$ git log -1
commit 12f0007ede35467a9c631dc8720961b77df53427 (HEAD -> master, origin/master, origin/HEAD)
```
When providing `auto_adjust_flush_quota: 2` in `scylla.yaml`, scylla operates normally and logs:
```
WARN  2023-12-13 13:06:15,601 [shard 0:main] init - Option is deprecated : auto_adjust_flush_quota
```
When providing e.g. `--auto-adjust-flush-quota on` via cmd line, scylla operates normally and logs:
```
WARN  2023-12-13 13:07:26,976 [shard 0:main] init - auto-adjust-flush-quota option ignored (deprecated)
```
so no issues here.

**Comment by kbr-scylla** [comment]
Do we plan to eventually turn `Deprecated` into `Unused` or will they stay `Deprecated` forever?

If we want to turn them, we probably should fix the bug -- someone may be starting Scylla and passing a Deprecated option by cmdline, which suddenly stops working as it's changed to Unused.

Definitely low priority though, they can always change the way they start Scylla.

**Comment by kostja** [comment]
Deprecated should be switched to Unused in the next minor. I agree with @ptrsmrn let's just follow the standard procedure even if the option is not used already - it doesn't break anything to first warn and only then ignore, this will give users a period to remove the option from the config file.
